### PR TITLE
Splitting Stereo Dataset PR(#6269)

### PR DESCRIPTION
**Pull request by TeodorPoncu**
Broken down PR(#6269) as per @NicolasHug [suggestion](https://github.com/pytorch/vision/pull/6269#pullrequestreview-1040016469)

### CODE REVIEW

File: torchvision/datasets/_stereo_matching.py
Comment (NicolasHug) [inline torchvision/datasets/_stereo_matching.py:161]: Should this just be `self._images = ...` ?

The `+=` suggests an update, but it's not really the case. Same everywhere else and for `self._disaprities` as well

File: torchvision/datasets/_stereo_matching.py
Comment (NicolasHug) [inline torchvision/datasets/_stereo_matching.py:62]: Not super important but I think you could be using `glob` from `Pathlib` instead of `glob.glob`, which would avoid having to convert all these `Path` to `str`.

File: test/datasets_utils.py
Comment (NicolasHug) [inline test/datasets_utils.py:754]: To avoid repeating code too much I'd suggest merging `shape_test_for_stereo_gt_w_mask` and `shape_test_for_stereo_gt_no_mask` and `shape_test_for_stereo_no_gt` together, possibly with new `no_gt` and `no_mask` boolean parameters

File: test/test_datasets.py
Comment (NicolasHug) [inline test/test_datasets.py:2748]: No need for a default here: it would be an error it the key is missing and we need to know about it

```suggestion
        num_examples = {"train": 4, "test": 6}[config["split"]]
```

File: test/test_datasets.py
Comment (NicolasHug) [inline test/test_datasets.py:2687]:
```suggestion
        num_examples = {"train": 4, "test": 3}[config["split"]]
```

**Review by NicolasHug — COMMENTED** [review]
Thanks @TeodorPoncu ,

I have a few comments but they're mostly minor or about conventions. The PR looks great in general, nice job!

Apart from the comments below, we'll need to add these classes to `datasets.rst` in order to look at the rendered docs.

File: torchvision/datasets/_stereo_matching.py
Comment (TeodorPoncu) [reply torchvision/datasets/_stereo_matching.py:115]: I actually mostly wrapped up writing the transform pipeline yesterday, and I have mixed feelings about the signature now. Since tuples are immutable, sometimes you've got to do a lot of unpacking and repacking inside the transform, so I'd actually be more inclined to use the flattened signature, even though it comes at the cost of increased verbosity.

Comment (datumbox) [reply torchvision/datasets/_stereo_matching.py:115]: @NicolasHug I'm missing context to provide an educated opinion. I understand that you are dealing with the legacy datasets and you have to "hack them together" to progress with the Depth Perception work, so in that sense I agree with your "keep as-is and re-evaluate" proposal.

I suspect that there are some incompatibilities with the current approach and the new Transforms API but that's not because you have multiple images/masks/maps but because left/right inputs need to be processed together rather than independently. I will probably need to understand more how this works on this task but it's an interesting use-case that we will need to support and might challenge some of the assumptions we made about the transforms and especially the fact the one that suggests that input can be transformed independently. cc @pmeier @vfdev-5

File: torchvision/datasets/_stereo_matching.py
Comment (TeodorPoncu) [reply torchvision/datasets/_stereo_matching.py:62]: As far as [I know](https://docs.python.org/3/library/pathlib.html) `Path.glob` requires the user to pass the **pattern** to be matched under the root specified by the `Path` object. Therefor for this change _scan_pairs would degenerate into something like `_scan_pairs(left_path, left_pattern, right_path, right_pattern)`.

File: torchvision/datasets/_stereo_matching.py
Comment (TeodorPoncu) [reply torchvision/datasets/_stereo_matching.py:309]: I would rather suggest to do `""` instead of `None`  as we can check with `if not file_path`. Setting it to a list of `None` would start throwing mypy errors for:

```python3
if paths_right_pattern:
            right_paths = sorted(glob(paths_right_pattern))
else:
            right_paths = list(None for _ in left_paths)
```

So, to address both this issue and https://github.com/pytorch/vision/pull/6311#discussion_r931259039 I would personally store `""` when no pattern is specified and switch to a `if not file_path` check.

File: test/test_datasets.py
Comment (TeodorPoncu) [reply test/test_datasets.py:2748]: Removing the `.get` when we're trying to test how the dataset responds when receiving a bad split name will throw an error: `FAILED test/test_datasets.py::Kitti2015StereoTestCase::test_bad_input - KeyError: 'bad'`

In the test case for `KittiFlow` it's a simple if else statement, so **unknown keys** will default to the else value. This sort of behaviour could not be achievable for datasets that have more than 2 possible split values.

https://github.com/pytorch/vision/blob/0ed5d81196868c0f87b7e3e89d28c998b759fc60/test/test_datasets.py#L1951

File: torchvision/datasets/_stereo_matching.py
Comment (NicolasHug) [reply torchvision/datasets/_stereo_matching.py:309]: > mypy errors

That's why in https://github.com/pytorch/vision/pull/6311/files#r931262310 I suggested to keep type annotations to the bare minimum (feel free to entirely remove them as well, I won't ask to bring them back).

As much as possible, we should avoid letting mypy be in the way of writing Pythonic code (it's a hard task, I would agree).

File: torchvision/datasets/_stereo_matching.py
Comment (TeodorPoncu) [reply torchvision/datasets/_stereo_matching.py:115]: To provide some context as to the specific use-case:

```python3
class RandomHorizontalFlip(torch.nn.Module):
    def __init__(self, p: float = 0.5) -> None:
        super().__init__()
        self.p = p

    def forward(
        self,
        images: Tuple[Tensor, Tensor],
        disparities: Tuple[Tensor, Tensor],
        masks: Tuple[Tensor, Tensor],
    ) -> Tuple[Tuple, Tuple, Tuple]:

        img_left, img_right = images
        dsp_left, dsp_right = disparities
        mask_left, mask_right = masks

        if dsp_right is not None and torch.rand(1) > self.p:
            # horizontal flip at pixel level
            img_left, img_right = F.hflip(img_left), F.hflip(img_right)
            dsp_left, dsp_right = F.hflip(dsp_left), F.hflip(dsp_right)
            # mask flipping flow is heavily skewed towards how we handle
            # generating pseudo-masks via the transform compose
            if mask_left is not None and mask_right is not None:
                mask_left, mask_right = F.hflip(mask_left), F.hflip(mask_right)
            # flip the two stereo channels
            return (
                (mask_right, mask_left),
                (dsp_right, dsp_left),
                (mask_right, mask_left)
            )

        return images, disparities, masks
```

The transform requires some inputs that for some datasets might not exist, namely the right disparity map. This is due to the fact, that we perform a horizontal pixel flip, the new target is not the transformed version of the original target but the opposite channel flipped target. Therefor the transform needs to have access to all the information.

Were it not for this transform, we could have just abided by the previous flow template.

Comment (TeodorPoncu) [reply torchvision/datasets/_stereo_matching.py:113]: For the bigger picture view, here is some extra context:

```python3
class Compose(torch.nn.Module):
    def __init__(self, transforms):
        super().__init__()
        self.transforms = transforms

    def forward(self, images, disparities, masks):
        for t in self.transforms:
            images, disparities, masks = t(images, disparities, masks)
        return images, disparities, masks
```

Flattening everything would require making transforms that support the following `forward` scheme:
```python3
    def forward(self, img_left, img_right, dsp_left, dsp_right, mask_left, mask_right):
        for t in self.transforms:
            img_left, img_right, dsp_left, dsp_right, mask_left, mask_right = t(img_left, img_right, dsp_left, dsp_right, mask_left, mask_right)
        return img_left, img_right, dsp_left, dsp_right, mask_left, mask_right
```

Comment (jdsgomes) [reply torchvision/datasets/_stereo_matching.py:113]: I am not very familiarised with the new transforms API, but looking into [it](https://github.com/pytorch/vision/blob/main/torchvision/prototype/transforms/_transform.py#L11) could we pass everything as a dictionary and return also a dictionary?

Comment (TeodorPoncu) [reply torchvision/datasets/_stereo_matching.py:113]: I'm not sure how much this is considered canonical workflow. Dicts however were my personal go-to choice when faced with not-so straight-forward data workflows.

```python3
stereo_dataset_inputs = {
    "left_image": ...,
    "right_image": ...,
    "left_disparity": ...,
    "right_disparity": ...,
    "left_mask": ...,
    "right_mask": ...,
}
stereo_dataset_inputs = transform(stereo_dataset_inputs)
return stereo_dataset_inputs["left_image"], ...
```

HuggingFace does something similar for [example](https://huggingface.co/datasets/wikitext)

Comment (datumbox) [reply torchvision/datasets/_stereo_matching.py:113]: Yes that kind of transforms are supported. But keep in mind that you will need to create custom classes that implement the logic the way you describe above. The API gives you goodies/utilities to assist your development.

Another thing you will need to do is possibly define a new custom type for Disparity Maps. We already offer support for images and masks and thankfully it's easy to define new types. There you would need to define what each operator can do to the maps and how they work (example: can you resize it using bilinear? how are colour shifts on images affect disparity maps? etc).

Comment (TeodorPoncu) [reply torchvision/datasets/_stereo_matching.py:113]: Truthfully, a **disparity map is actually just a flow map** with a single channel / **0-flow on the y-axis**. The same constraints that apply to the transforms in our optical flow reference scripts apply to these as well.

Bilinear resizing will still work as intended because both flows and disparities are measured in pixels and not some other real-world distance metric.

Judging by what several other paper repos provide in terms of augmentation pipelines ([1](https://github.com/megvii-research/CREStereo/blob/master/dataset.py) , [2](https://github.com/princeton-vl/RAFT-Stereo/blob/main/core/stereo_datasets.py)) it seems that the same approach is being followed.

Typically in monocular depth-estimation the same problems are posed and although [lambertian surfaces](https://en.wikipedia.org/wiki/Lambertian_reflectance) are an important issue, the assumption that color shifts should affect the depth-map are not taken into account. Both in the unsupervised and supervised setting we consider such alterations something that should not affect the **true world** model / estimation, but they are solely camera related / a regularization method.

Comment (jdsgomes) [reply torchvision/datasets/_stereo_matching.py:113]: In any case I agree that we can merge this PR to unblock and re-visit this considering the new transforms API.

Comment (datumbox) [reply torchvision/datasets/_stereo_matching.py:113]: OK sounds like the new API in principle supports both the multiple inputs (named or not) and special types like disparity maps. I don't see immediately an incompatibility but we should keep an eye for similar requests as this is a useful feedback for the new API.

Comment (TeodorPoncu) [reply torchvision/datasets/_stereo_matching.py:113]: @datumbox @NicolasHug since, since this PR is non-blocking for the model training, I could try dog-fooding the new transforms API and include in this PR parts of the reference script in order for us to get a clearer idea how everything would mix in with the new API and have a better picture of how the dataset should be structured. 

Is that ok with you guys?

Comment (datumbox) [reply torchvision/datasets/_stereo_matching.py:113]: We literally just merged a bunch of updates today at #6305. It might be a bit premature to dogfood it (we'll definitely take upon your offer a bit later though) but you can already have a look how it works on main branch at `torchvision.prototype.transforms`.

**Review by jdsgomes — APPROVED** [review]
LGTM. Regarding the open discussion I think we agree that this can be merged and revisited.  I would just wait for @NicolasHug  approval since he made most comments/suggestions.

File: torchvision/datasets/_stereo_matching.py
Comment (NicolasHug) [inline torchvision/datasets/_stereo_matching.py:49]: Do we really need to add `type: ignore` here? I'd suggest to type the rest, or not type anything at all (which is by far my prefered way)

File: torchvision/datasets/_stereo_matching.py
Comment (NicolasHug) [inline torchvision/datasets/_stereo_matching.py:64]: This should do:

```suggestion
        left_paths = list(sorted(glob(paths_left_pattern)))

        right_paths: List[Union[None, str]]
        if paths_right_pattern:
            right_paths = list(sorted(glob(paths_right_pattern)))
        else:
            right_paths = list(None for _ in left_paths)
```

**Review by NicolasHug — COMMENTED** [review]
Thanks for addressing the comments @TeodorPoncu , I made another pass below


Would you mind adding the datasets to `datasets.rst` so that we can check the rendered docs as well? Thanks!

File: torchvision/datasets/_stereo_matching.py
Comment (TeodorPoncu) [reply torchvision/datasets/_stereo_matching.py:49]: I've left the `#type: ingore` because mypy starts throwing errors about empty lists since it cannot infer their dtype.

`torchvision/datasets/_stereo_matching.py:48: error: Need type annotation for "_images" (hint: "_images: List[<type>] = ...")  [var-annotated]
            self._images = []
`

`torchvision/datasets/_stereo_matching.py:49: error: Need type annotation for "_disparities" (hint: "_disparities: List[<type>] = ...")  [var-annotated]
            self._disparities = []
`

File: torchvision/datasets/_stereo_matching.py
Comment (TeodorPoncu) [reply torchvision/datasets/_stereo_matching.py:64]: Worked like a charm! Thank you very much!

File: torchvision/datasets/_stereo_matching.py
Comment (NicolasHug) [inline torchvision/datasets/_stereo_matching.py:260]: This does not render properly in the docs: https://output.circle-artifacts.com/output/job/4dcd20b4-ad55-48a3-a0c9-474bfaae3aa3/artifacts/0/docs/generated/torchvision.datasets.Kitti2015Stereo.html#torchvision.datasets.Kitti2015Stereo.__getitem__

I think you just need to skip lines like so:

```
        """Return example at given index.

        Args:
            index(int): The index of the example to retrieve

        Returns:
```

Same for the other datasets :)

**Review by NicolasHug — APPROVED** [review]
Thanks @TeodorPoncu , one last cosmetic comment regarding the dogs but other than that LGTM! I'll approve now so that you can merge when it's addressed. Nice work!

File: torchvision/datasets/_stereo_matching.py
Comment (NicolasHug) [inline torchvision/datasets/_stereo_matching.py:177]: The line skip should be between these two as well:

```suggestion
        """Return example at given index.
        
        Args:
```

**Pull request merged**
